maemo-launcher/launcher/launcher.1.in
.\" $Id$
.TH maemo\-launcher 1 2006-09-13 @VERSION@ "Maemo Launcher manual"
.SH NAME
maemo\-launcher \- initialization launcher daemon
.SH SYNOPSIS
.B maemo\-launcher
.RI [ options ]
.SH DESCRIPTION
.B maemo\-launcher
is an initialization speedup daemon, that tries to move as much as possible
of tasks done for every application to a central program that then forks
and dynamically opens the applications to run.
.PP
The application to run is communicated from the
.B maemo\-invoker
through a unix socket and then started.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
.B \-\-daemon
Fork and go to background.
.TP
.B \-\-pidfile \fIPIDFILE\fP
Specify a different filename for the pid file.
.TP
.BR \-\-booster " \fIMODULE\fP,\fI...\fP"
Specify a booster module to use, to preinitialize some state.
.TP
.B \-\-quiet
Do not print anything.
.TP
.B \-\-help
Show summary of options.
.TP
.B \-\-version
Show version of program.
.SH SEE ALSO
.BR maemo\-invoker (1).
.SH AUTHOR
.B maemo\-launcher
was written by Michael Natterer <mitch@imendio.com> and
Guillem Jover <guillem.jover@nokia.com>.


ChangeLog
2007-03-15  Xan Lopez  <xan.lopez@nokia.com>

	* gtk+-2.0.pc.in (Cflags): remove -DMAEMO_CHANGES, it's intentionally
	only in hildon.

2007-03-14  Michael Natterer  <mitch@imendio.com>

	* gtk+-2.0.pc (Cflags): add -DMAEMO_CHANGES so library users see
	the changes.

2007-03-13  Xan Lopez  <xan.lopez@nokia.com>

	* gtk/gtkentry.h:
	* gtk/gtkimcontext.h:
	* gtk/gtktreeview.h:

	Restore the upstream ABI compatibility.
	
2007-03-13  Xan Lopez  <xan.lopez@nokia.com>

	* gtk/gtkenums.h:
	* gtk/gtkiconfactory.c:
	* gtk/gtkstyle.c:

	Remove the HILDON_ICON_SIZE_* stuff from the stock enums.

2007-03-13  Michael Natterer  <mitch@imendio.com>

	* gdk/x11/gdkdisplay-x11.c (gdk_display_open): added missing
	#ifdef MAEMO_CHANGES.

2007-03-13  Michael Natterer  <mitch@imendio.com>

	* gdk/gdkwindow.h (enum GdkWindowTypeHint): remove
	GDK_WINDOW_TYPE_HINT_MESSAGE.

	* gdk/x11/gdkwindow-x11.c (gdk_window_set_type_hint)
	(gdk_window_get_type_hint): remove compat support for
	TYPE_HINT_MESSAGE.

	* gtk/gtkmenu.c (gtk_menu_window_visibility_notify_event): use
	TYPE_HINT_NOTIFICATION instead of TYPE_HINT_MESSAGE.

2007-03-13  Xan Lopez  <xan.lopez@nokia.com>

	* gtk/gtkwindow.h:

	Remove the focus_prev API from the header file.
	
2007-03-13  Xan Lopez  <xan.lopez@nokia.com>

	* gtk/gtkwidget.h:

	Remove the focus handling API from the header file.
	
2007-03-13  Xan Lopez  <xan.lopez@nokia.com>

	* gtk/gtktextbuffer.h:

	Remove the old rich text copy&paste API from the header.
	
2007-03-13  Xan Lopez  <xan.lopez@nokia.com>

	* gtk/gtkbutton.c:

	We really need the type to compile, genius.
	
2007-03-13  Xan Lopez  <xan.lopez@nokia.com>

	* gtk/gtkrange.h:

	Remove osso_gtk_range_* API from the header file.
	
2007-03-13  Xan Lopez  <xan.lopez@nokia.com>

	* gtk/gtkfilesystem.h:

	Remove hildon_gtk_file_system_create and
	gtk_file_system_render_icon from the header file.
	
2007-03-13  Xan Lopez  <xan.lopez@nokia.com>

	* gtk/gtkbutton.[ch]:

	Remove hildon_gtk_button_set_depressed,
	osso_gtk_button_set_detail_from_attach_flags and the
	OssoGtkButtonAttachFlags type from the header
	files.
	Hardcode the number of items of the enum in the
	C file so we don't depend on its definition. Yay hacks.
	
2007-03-13  Xan Lopez  <xan.lopez@nokia.com>

	* gtk/gtkrc.c:
	* gtk/gtkstyle.h:

	Remove support for logical colors with the maemo gtk+ 2.6 format.

2007-03-08  Xan Lopez  <xan.lopez@nokia.com>

	* gtk/gtkentry.c: Remove the icon-width property, it's unused.

2007-03-05  Xan Lopez  <xan.lopez@nokia.com>

	* gtk/gtkicontheme.c: Back out our changes to the icon loading order.
	This should fix stock icons being broken and has no noticeable performance
	differences in out tests.

2007-02-21  Lucas Rocha  <lucas.rocha@nokia.com>

	* gtk/gtkiconview.c: surround all changes made on selection behavior
	with MAEMO_CHANGES ifdefs.

Wed Feb 21 13:43:33 2007  Tim Janik  <timj@gtk.org>

	* gtkprogressbar.c: adapt minimum progressbar height (used for progress
	bars without text) to maemo theme image height.

2007-02-20  Lucas Rocha  <lucas.rocha@nokia.com>

	* gtk/gtkiconview.c: selection painted around the item, not just the
	text.
	(gtk_icon_view_paint_item): paint selection around the role item
	instead of just around the text cell. If an editable/activatable 
	cell is found, draw focus on it. If none is found, draw focus  
	around the whole item.
	(gtk_icon_view_button_press): handle button press considering button
	press inside item's area, not only on item cells.
	(gtk_icon_view_get_path_at_pos): consider the whole item area, not
	only its cells.

2007-02-20  Michael Natterer  <mitch@imendio.com>

	* gtk/gtkcombobox.c (gtk_combo_box_list_key_press): pop down
	on F4 (= menu) and F5 (= home) (can't be done via gtkrc).

2007-02-20  Kristian Rietveld  <kris@imendio.com>

	* gtk/gtkcellrenderertoggle.c: back out checkbox-mode.

2007-02-20  Kristian Rietveld  <kris@imendio.com>

	* gdk/gdkwindow.h: modify GdkWindowTypeHint so that we don't end
	the enumerator list with a comma.  (MB#1087).

2007-02-20  Michael Natterer  <mitch@imendio.com>

	* gtk/gtkmenuitem.c (_gtk_menu_item_popup_submenu): queue a draw
	on the menu item so the changed arrow shadow becomes visible.

2007-02-20  Kristian Rietveld  <kris@imendio.com>

	Merge from upstream trunk:

    2006-12-15  Dom Lachowicz <domlachowicz@gmail.com>

	* gtk/gtkcombobox.c: Make GtkComboBox in "appears-as-list" mode
	more closely resemble its Win32 counterpart. Bug #340204.

2007-02-20  Michael Natterer  <mitch@imendio.com>

	* gtk/gtkmenuitem.c (gtk_menu_item_paint): clean up yesterday's
	commit: change HILDON_ARROW_SPACE of 6 pixels to ARROW_PADDING of
	4 pixels because that's how it's used. Use GTK_STATE_NORMAL
	instead of 0. Cleaned up a bit and added comments.

2007-02-19  Michael Natterer  <mitch@imendio.com>

	* gtk/gtkmenuitem.c (gtk_menu_item_paint): port over changes to
	menu item drawing when there is a submenu.

2007-02-16  Michael Natterer  <mitch@imendio.com>

	* gtk/gtkcombobox.c (gtk_combo_box_menu_setup)
	(gtk_combo_box_list_setup): if the child widget is a GtkEntry,
	unset CAN_FOCUS on the toggle button.

	(gtk_combo_box_list_button_pressed): grab the focus on the combo
	box, which implicitly grabs the right widget due to grab_focus()
	virtualization.

2007-02-16  Tommi Komulainen  <tommi.komulainen@nokia.com>

	* gtk/gtkwidget.c (gtk_widget_realize): Call
	gtk_widget_set_extension_events_internal even with !NO_WINDOW - it
	does the right checks to ensure it only touches windows belonging to
	the widget. MB#1060

Thu Feb 15 15:18:37 2007  Tim Janik  <timj@gtk.org>

	* gtk/gtkdialog.c: added style property ::content-area-spacing, to
	allow configurable area spacing similar to the style properties 
	::button-spacing and ::content-area-border, upstream #408244.

Thu Feb 15 13:17:52 2007  Tim Janik  <timj@gtk.org>

	* gtk/gtkarrow.c (gtk_arrow_expose): removed arrow size down scaling
	to 70% to allow full arrow themability in maemo.


maemo-launcher/launcher/launcher.c
/*
 * $Id$
 *
 * Copyright (C) 2005, 2006 Nokia Corporation
 *
 * Authors: Michael Natterer <mitch@imendio.com>
 *	    Guillem Jover <guillem.jover@nokia.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
 *
 */

#define _GNU_SOURCE

#include "config.h"

#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/resource.h>
#include <fcntl.h>
#include <signal.h>
#include <unistd.h>
#include <errno.h>

#include "report.h"
#include "invokelib.h"
#include "comm_dbus.h"
#include "booster.h"
#include "prog.h"

/* FIXME: Should go into '/var/run/'. */
#define LAUNCHER_PIDFILE "/tmp/"PROG_NAME".pid"

typedef struct
{
  char *name;
  pid_t pid;
  int sock;
} child_t;

typedef struct
{
  int n;
  int used;
  child_t *list;
} kindergarten_t;

static char *pidfilename = LAUNCHER_PIDFILE;
static pid_t is_parent = 1;
static volatile bool sigchild_catched = false;
static volatile bool sighup_catched = false;
static bool send_app_died = false;

#define OOM_ENABLE "0"

static void
oom_unprotect(void)
{
  const char *filename = "/proc/self/oom_adj";
  int fd;

  fd = open(filename, O_WRONLY);
  if (fd < 0)
  {
    error("opening file '%s'\n", filename);
    return;
  }

  if (write(fd, OOM_ENABLE, strlen(OOM_ENABLE)) < 0)
  {
    if (errno == EPERM)
      error("kernel w/o support for user processes rising the oom value\n");
    else
      error("could not write to file '%s'\n", filename);
  }

  close(fd);
}

static bool
rise_oom_defense(pid_t pid)
{
  pid_t defender_pid;
  int status;

  defender_pid = fork();
  if (defender_pid)
    waitpid(defender_pid, &status, 0);
  else
  {
    char pidstr[20];

    snprintf(pidstr, sizeof(pidstr), "%d", pid);
    execl(DEFENDER, DEFENDER, pidstr, NULL);
    _exit(1);
  }

  if (WIFEXITED(status) && WEXITSTATUS(status) == 0)
    return true;
  else
  {
    error("rising the oom shield for pid=%d status=%d\n", pid, status);
    return false;
  }
}

static void
launch_process(prog_t *prog)
{
  int cur_prio;

  load_main(prog);

  /* Possibly restore process priority. */
  errno = 0;
  cur_prio = getpriority(PRIO_PROCESS, 0);
  if (!errno && cur_prio < prog->prio)
    setpriority(PRIO_PROCESS, 0, prog->prio);

  /* Unprotect our worthless childs from the oom monster. */
  if (prog->prio >= 0)
    oom_unprotect();

  print_prog_env_argv(prog);

  debug("launching process: '%s'\n", prog->filename);

#ifdef DEBUG
  report_set_output(report_console);
#else
  report_set_output(report_none);
#endif

  _exit(prog->entry(prog->argc, prog->argv));
}

static int
invoked_init(void)
{
  int fd;
  struct sockaddr_un sun;

  fd = socket(PF_UNIX, SOCK_STREAM, 0);
  if (fd < 0)
    die(1, "opening invoker socket\n");

  unlink(INVOKER_SOCK);
  umask(022);

  sun.sun_family = AF_UNIX;
  strcpy(sun.sun_path, INVOKER_SOCK);

  if (bind(fd, &sun, sizeof(sun)) < 0)
    die(1, "binding to invoker socket\n");

  if (listen(fd, 10) < 0)
    die(1, "listening to invoker socket\n");

  return fd;
}

static bool
invoked_get_magic(int fd, prog_t *prog)
{
  uint32_t msg;

  /* Receive the magic. */
  invoke_recv_msg(fd, &msg);
  if ((msg & INVOKER_MSG_MASK) == INVOKER_MSG_MAGIC)
  {
    if ((msg & INVOKER_MSG_MAGIC_VERSION_MASK) == INVOKER_MSG_MAGIC_VERSION)
      invoke_send_msg(fd, INVOKER_MSG_ACK);
    else
    {
      error("receiving bad magic version (%08x)\n", msg);
      return false;
    }
  }
  else
  {
    error("receiving bad magic (%08x)\n", msg);
    return false;
  }

  prog->options = msg & INVOKER_MSG_MAGIC_OPTION_MASK;

  return true;
}

static bool
invoked_get_name(int fd, prog_t *prog)
{
  uint32_t msg;

  /* Get the action. */
  invoke_recv_msg(fd, &msg);
  if (msg != INVOKER_MSG_NAME)
  {
    error("receiving invalid action (%08x)\n", msg);
    return false;
  }

  prog->name = invoke_recv_str(fd);
  if (!prog->name)
    return false;

  invoke_send_msg(fd, INVOKER_MSG_ACK);

  return true;
}

static bool
invoked_get_exec(int fd, prog_t *prog)
{
  prog->filename = invoke_recv_str(fd);
  if (!prog->filename)
    return false;

  invoke_send_msg(fd, INVOKER_MSG_ACK);

  return true;
}

static bool
invoked_get_args(int fd, prog_t *prog)
{
  int i;
  uint32_t msg;
  size_t size;

  /* Get argc. */
  invoke_recv_msg(fd, &msg);
  prog->argc = msg;
  size = msg * sizeof(char *);
  if (size < msg)
  {
    error("on buggy or malicious invoker code, heap overflow avoided\n");
    return false;
  }
  prog->argv = malloc(size);
  if (!prog->argv)
  {
    error("mallocing\n");
    return false;
  }

  /* Get argv. */
  for (i = 0; i < prog->argc; i++)
  {
    prog->argv[i] = invoke_recv_str(fd);
    if (!prog->argv[i])
    {
      error("receiving argv[%i]\n", i);
      return false;
    }
  }

  invoke_send_msg(fd, INVOKER_MSG_ACK);

  return true;
}

static bool
invoked_get_prio(int fd, prog_t *prog)
{
  uint32_t msg;

  invoke_recv_msg(fd, &msg);
  prog->prio = msg;

  invoke_send_msg(fd, INVOKER_MSG_ACK);

  return true;
}

static bool
invoked_send_pid(int fd, pid_t pid)
{
  invoke_send_msg(fd, INVOKER_MSG_PID);
  invoke_send_msg(fd, pid);

  return true;
}

static bool
invoked_get_actions(int fd, prog_t *prog)
{
  while (1)
  {
    uint32_t msg;

    /* Get the action. */
    invoke_recv_msg(fd, &msg);

    switch (msg)
    {
    case INVOKER_MSG_EXEC:
      invoked_get_exec(fd, prog);
      break;
    case INVOKER_MSG_ARGS:
      invoked_get_args(fd, prog);
      break;
    case INVOKER_MSG_PRIO:
      invoked_get_prio(fd, prog);
      break;
    case INVOKER_MSG_END:
      invoke_send_msg(fd, INVOKER_MSG_ACK);

      if (prog->options & INVOKER_MSG_MAGIC_OPTION_WAIT)
        invoked_send_pid(fd, getpid());

      return true;
    default:
      error("receiving invalid action (%08x)\n", msg);
      return false;
    }
  }
}

static bool
invoked_send_exit(int fd, int status)
{
  invoke_send_msg(fd, INVOKER_MSG_EXIT);
  invoke_send_msg(fd, status);

  return true;
}

static void
clean_daemon(int signal)
{
  if (is_parent)
  {
    unlink(pidfilename);
    unlink(INVOKER_SOCK);
    killpg(0, signal);
  }

  exit(0);
}

static kindergarten_t *
alloc_childs(int n)
{
  kindergarten_t *childs;

  childs = malloc(sizeof(kindergarten_t));
  if (!childs)
  {
    error("allocating a kindergarten\n");
    return NULL;
  }

  childs->used = 0;
  childs->n = n;
  childs->list = calloc(n, sizeof(child_t));
  if (!childs->list)
  {
    error("allocating a child list\n");
    return NULL;
  }

  return childs;
}

static bool
grow_childs(kindergarten_t *childs)
{
  int halfsize = childs->n * sizeof(child_t);
  int size = halfsize * 2;
  void *p;

  p = realloc(childs->list, size);
  if (!p)
    return false;

  childs->list = p;
  memset(childs->list + childs->n, 0, halfsize);

  childs->n *= 2;

  return true;
}

static bool
release_childs(kindergarten_t *childs)
{
  int i;
  child_t *list = childs->list;

  for (i = 0; i < childs->used; i++)
  {
    close(list[i].sock);
    free(list[i].name);
  }

  childs->used = 0;
  childs->n = 0;
  free(childs->list);

  return true;
}

static int
get_child_slot_by_pid(kindergarten_t *childs, pid_t pid)
{
  child_t *list = childs->list;
  int i;

  for (i = 0; i < childs->n; i++)
    if (list[i].pid == pid)
      return i;

  return -1;
}

static bool
invoked_send_fake_exit(int sock)
{
  /* Send a fake exit code, so the invoker does not wait for us. */
  invoked_send_exit(sock, 0);
  close(sock);

  return true;
}

static bool
assign_child_slot(kindergarten_t *childs, child_t *child)
{
  int id;

  if (childs->used == childs->n && !grow_childs(childs))
  {
    error("cannot make a bigger kindergarten, not tracking child %d\n",
	  child->pid);
    invoked_send_fake_exit(child->sock);
    return false;
  }

  id = get_child_slot_by_pid(childs, 0);
  if (id < 0)
  {
    error("this cannot be happening! no free slots on the kindergarten,\n"
	  "not tracking child %d\n", child->pid);
    invoked_send_fake_exit(child->sock);
    return false;
  }

  childs->list[id].name = child->name;
  childs->list[id].sock = child->sock;
  childs->list[id].pid = child->pid;
  childs->used++;

  return true;
}

static bool
child_died_painfully(int status)
{
  if ((WIFEXITED(status) && WEXITSTATUS(status) != 0) ||
      (WIFSIGNALED(status) && (WTERMSIG(status) != SIGINT &&
			       WTERMSIG(status) != SIGTERM &&
			       WTERMSIG(status) != SIGKILL)))
    return true;
  else
    return false;
}

static bool
release_child_slot(kindergarten_t *childs, pid_t pid, int status)
{
  int id = get_child_slot_by_pid(childs, pid);

  if (id >= 0)
  {
    child_t *child = &childs->list[id];

    if (send_app_died && child_died_painfully(status))
      comm_send_app_died(child->name, pid, status);

    invoked_send_exit(child->sock, status);

    close(child->sock);
    free(child->name);

    child->name = NULL;
    child->sock = 0;
    child->pid = 0;

    childs->used--;
  }
  else
    info("no child %i found in the kindergarten.\n", pid);

  return true;
}

static void
clean_childs(kindergarten_t *childs)
{
  int status;
  pid_t childpid;

  while ((childpid = waitpid(-1, &status, WNOHANG)) > 0)
  {
    release_child_slot(childs, childpid, status);

    if (WIFEXITED(status))
      info("child (pid=%d) terminated due to exit()=%d\n", childpid,
	   WEXITSTATUS(status));
    else if (WIFSIGNALED(status))
      info("child (pid=%d) terminated due to signal=%d\n", childpid,
	   WTERMSIG(status));
    else
      info("child (pid=%d) terminated due to unknown status=%d\n", status);
  }
}

static void
sigchild_handler(int sig)
{
  sigchild_catched = sig;
}

static void
sighup_handler(int sig)
{
  sighup_catched = true;
}

static void
sigs_init(void)
{
  struct sigaction sig;

  memset(&sig, 0, sizeof(sig));
  sig.sa_flags = SA_RESTART;

  sig.sa_handler = SIG_IGN;
  sigaction(SIGPIPE, &sig, NULL);

  sig.sa_handler = clean_daemon;
  sigaction(SIGINT, &sig, NULL);
  sigaction(SIGTERM, &sig, NULL);

  sig.sa_handler = sigchild_handler;
  sigaction(SIGCHLD, &sig, NULL);

  sig.sa_handler = sighup_handler;
  sigaction(SIGHUP, &sig, NULL);
}

static void
sigs_restore(void)
{
  struct sigaction sig;

  memset(&sig, 0, sizeof(sig));
  sig.sa_handler = SIG_DFL;
  sig.sa_flags = SA_RESTART;

  sigaction(SIGPIPE, &sig, NULL);
  sigaction(SIGINT, &sig, NULL);
  sigaction(SIGTERM, &sig, NULL);
  sigaction(SIGCHLD, &sig, NULL);
  sigaction(SIGHUP, &sig, NULL);
}

static void
sigs_interrupt(int flag)
{
  siginterrupt(SIGCHLD, flag);
}

static void
env_init(void)
{
  unsetenv("LD_BIND_NOW");
}

static void
create_pidfile(void)
{
  FILE *pidfile = fopen(pidfilename, "w");

  if (pidfile)
  {
    pid_t pid = getpid();
    fprintf(pidfile, "%d\n", pid);
    fclose(pidfile);
  }
}

static void
daemonize(void)
{
  pid_t pid;

  if (fork())
    _exit(0);

  if (setsid() < 0)
    die(1, "setting session id");

  pid = fork();
  if (pid < 0)
    die(1, "forking while daemonizing");
  else if (pid)
    _exit(0);

  chdir("/");

  /* Protect us from the oom monster. */
  rise_oom_defense(getpid());
}

static void
console_quiet(void)
{
  report_set_output(report_syslog);

  close(0);
  close(1);
  close(2);

  if (open("/dev/null", O_RDONLY) < 0)
    die(1, "opening /dev/null readonly");
  if (dup(open("/dev/null", O_WRONLY)) < 0)
    die(1, "opening /dev/null writeonly");
}

static void
version(void)
{
  printf("%s (%s) %s\n", PROG_NAME, PACKAGE, VERSION);

  exit(0);
}

static void
usage(int status)
{
  printf("Usage: %s [options]\n"
	 "\n"
	 "Options:\n"
	 "  --daemon            Fork and go into the background.\n"
	 "  --pidfile FILE      Specify a different pid file (default %s).\n"
	 "  --booster MODULE,...\n"
	 "                      Specify a comma-separated list of boosters to use"
	 "                      (default '%s').\n"
	 "  --send-app-died     Send application died signal.\n"
	 "  --quiet             Do not print anything.\n"
	 "  --version           Print program version.\n"
	 "  --help              Print this help message.\n"
	 "\n"
	 "Use the invoker to start a <shared object> from the launcher.\n"
	 "Where <shared object> is a binary including a 'main' symbol.\n"
	 "Note that the binary needs to be linked with -shared or -pie.\n",
	 PROG_NAME, LAUNCHER_PIDFILE, BOOSTER_DEFAULT);

  exit(status);
}

int
main(int argc, char *argv[])
{
  booster_t *boosters = NULL;
  kindergarten_t *childs;
  const int initial_child_slots = 64;
  int i;
  int fd;
  bool daemon = false;
  bool quiet = false;

  /*
   * Parse arguments.
   */
  for (i = 1; i < argc; ++i)
  {
    if (strcmp(argv[i], "--daemon") == 0)
      daemon = true;
    else if (strcmp(argv[i], "--quiet") == 0)
      quiet = true;
    else if (strcmp(argv[i], "--pidfile") == 0)
    {
      if (argv[++i])
	pidfilename = argv[i];
    }
    else if (strcmp(argv[i], "--booster") == 0)
    {
      if (argv[++i])
	boosters_alloc(&boosters, argv[i]);
    }
    else if (strcmp(argv[i], "--send-app-died") == 0)
      send_app_died = true;
    else if (strcmp(argv[i], "--version") == 0)
      version();
    else if (strcmp(argv[i], "--help") == 0)
      usage(0);
    else
      usage(-1);
  }

  if (!boosters)
    boosters_alloc(&boosters, BOOSTER_DEFAULT);

  /*
   * Daemon initialization.
   */
  boosters_load(boosters, &argc, &argv);

  sigs_init();
  env_init();

  /* Setup child tracking. */
  childs = alloc_childs(initial_child_slots);

  /* Setup the conversation channel with the invoker. */
  fd = invoked_init();

  if (daemon)
    daemonize();

  if (quiet)
    console_quiet();

  create_pidfile();

  /*
   * Application invokation loop.
   */
  while (1)
  {
    prog_t prog;
    int sd;
    int errno_accept;

    /* Accept a new invokation. */
    sigs_interrupt(1);
    sd = accept(fd, NULL, NULL);
    errno_accept = errno;
    sigs_interrupt(0);

    /* Handle signals received. */
    if (sigchild_catched)
    {
      clean_childs(childs);
      sigchild_catched = false;
    }

    if (sighup_catched)
    {
      boosters_reload(boosters);
      sighup_catched = false;
    }

    /* Minimal error handling. */
    if (sd < 0)
    {
      if (errno_accept != EINTR)
	error("accepting connections (%s)\n", strerror(errno_accept));

      continue;
    }

    /* Start conversation with the invoker. */
    memset(&prog, 0, sizeof(prog));
    if (!invoked_get_magic(sd, &prog))
    {
      close(sd);
      continue;
    }
    if (!invoked_get_name(sd, &prog))
    {
      close(sd);
      continue;
    }

    is_parent = fork();
    switch (is_parent)
    {
    case -1: /* Error. */
      error("forking while invoking\n");
      break;

    case 0: /* Child. */
      if (setsid() < 0)
	error("setting session id\n");

      sigs_restore();

      invoked_get_actions(sd, &prog);

      close(sd);
      close(fd);
      free(prog.name);

      release_childs(childs);

      /* Invoke it. */
      if (prog.filename)
      {
        info("invoking '%s'\n", prog.filename);
        set_progname(prog.argv[0], argc, argv);

	boosters_init(boosters, prog.argv[0]);

	launch_process(&prog);
      }
      else
        error("nothing to invoke\n");

      _exit (0);
      break;

    default: /* Parent. */
      if (prog.options & INVOKER_MSG_MAGIC_OPTION_WAIT)
      {
	child_t child;

	child.pid = is_parent;
	child.sock = sd;
	child.name = prog.name;

	assign_child_slot(childs, &child);
      }
      else
	close(sd);
      break;
    }
  }

  return 0;
}

